### PR TITLE
Add the correct nullable annotations to generated iterator code

### LINKED ISSUE #30010: Nullable attribute is not synthesized for the getter of the Current property emitted to "implement" iterator method
```
        [Fact]
        public void EmitAttribute_Iterator_01()
        {
            var source =
@"using System.Collections.Generic;
class C
{
    static IEnumerable<object?> F()
    {
        yield break;
    }
}";
            CompileAndVerify(
                source,
                parseOptions: TestOptions.Regular8,
                options: TestOptions.DebugDll.WithMetadataImportOptions(MetadataImportOptions.All),
                symbolValidator: module =>
                {
                    var property = module.ContainingAssembly.GetTypeByMetadataName("C").GetTypeMember("<F>d__0").GetProperty("System.Collections.Generic.IEnumerator<System.Object>.Current");
                    AssertNoNullableAttribute(property.GetAttributes());
                    var method = property.GetMethod;
                    // PROTOTYPE(NullableReferenceTypes): No synthesized attributes for this
                    // case which is inconsisten with IEnumerable<object?[]> in test below.
                    AssertNoNullableAttribute(method.GetReturnTypeAttributes());
                    AssertAttributes(method.GetAttributes(), "System.Diagnostics.DebuggerHiddenAttribute");
                });
        }
```

### PR DESCRIPTION
**Pull request by chsienki**
This is mainly about plumbing through `TypeSymbolWithAnnotations` to the correct point so we can generate the attribute based on it's value.

- Expose the `IteratorElementType` as a `TypeSymbolWithAnnotations`
- Return the annotated type, or a default as needed from the required locations
- Allow state machine field to take a TSWA and pass it in via the iterator rewriter
- Fix test

Fixes #30010

### CODE REVIEW

**Comment by chsienki** [comment]
@dotnet/roslyn-compiler for review please

File: src/Compilers/CSharp/Portable/Lowering/MethodToClassRewriter.cs [hunk 1]
```diff
@@ -672,12 +672,12 @@ internal override void AddSynthesizedAttributes(PEModuleBuilder moduleBuilder, r
...
             }
 
-            internal override TypeSymbol IteratorElementType
+            internal override TypeSymbolWithAnnotations IteratorElementType
             {
                 get
                 {
                     // BaseMethodWrapperSymbol should not be rewritten by the IteratorRewriter
-                    return null;
+                    return TypeSymbolWithAnnotations.Create(null);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Lowering/MethodToClassRewriter.cs]: >TypeSymbolWithAnnotations.Create(null) [](start = 27, length = 38)

`default`

File: src/Compilers/CSharp/Portable/Symbols/MethodSymbol.cs [hunk 2]
```diff
@@ -900,17 +900,17 @@ internal bool IsIterator
...
         /// <summary>
         /// If the method was written as an iterator method (i.e. with yield statements in its body) returns the
-        /// element type of the iterator.  Otherwise returns null.
+        /// element type of the iterator.  Otherwise returns default(TypeSymbolWithAnnotations).
         /// </summary>
-        internal virtual TypeSymbol IteratorElementType
+        internal virtual TypeSymbolWithAnnotations IteratorElementType
         {
-            get { return null; }
+            get { return TypeSymbolWithAnnotations.Create(null); }
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/MethodSymbol.cs]: >TypeSymbolWithAnnotations.Create(null) [](start = 25, length = 38)

`default`

File: src/Compilers/CSharp/Portable/Symbols/Source/LocalFunctionSymbol.cs [hunk 3]
```diff
@@ -289,16 +289,16 @@ public override bool IsExtensionMethod
...
             get
             {
                 return _iteratorElementType;
             }
             set
             {
-                Debug.Assert((object)_iteratorElementType == null || TypeSymbol.Equals(_iteratorElementType, value, TypeCompareKind.ConsiderEverything2));
-                Interlocked.CompareExchange(ref _iteratorElementType, value, null);
+                Debug.Assert(_iteratorElementType.IsDefault || TypeSymbol.Equals(_iteratorElementType.TypeSymbol, value.TypeSymbol, TypeCompareKind.ConsiderEverything2));
+                _iteratorElementType = value;
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Source/LocalFunctionSymbol.cs]: >_iteratorElementType = value; [](start = 16, length = 29)

The assignment is not thread-safe.

`_iteratorElementType` should probably be a `TypeSymbolWithAnnotations.Builder`. 

Then this property would be:
```
get => _iteratorElementType.ToType();
set => _iteratorElementType.InterlockedInitialize(value);
```

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberMethodSymbol.cs [hunk 4]
```diff
@@ -706,16 +706,16 @@ internal sealed override bool TryGetThisParameter(out ParameterSymbol thisParame
...
             get
             {
                 return _iteratorElementType;
             }
             set
             {
-                Debug.Assert((object)_iteratorElementType == null || TypeSymbol.Equals(_iteratorElementType, value, TypeCompareKind.ConsiderEverything2));
-                Interlocked.CompareExchange(ref _iteratorElementType, value, null);
+                Debug.Assert(_iteratorElementType.IsDefault || TypeSymbol.Equals(_iteratorElementType.TypeSymbol, value.TypeSymbol, TypeCompareKind.ConsiderEverything2));
+                _iteratorElementType = value;
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberMethodSymbol.cs]: >_iteratorElementType = value [](start = 16, length = 28)

The assignment is not thread-safe.

File: src/Compilers/CSharp/Portable/Lowering/MethodToClassRewriter.cs (hunk 1 above)
Comment (jaredpar) [reply src/Compilers/CSharp/Portable/Lowering/MethodToClassRewriter.cs]: Rather that `default` should we add a `static` property with a decent name: Missing, Null, etc ...

**Comment by chsienki** [comment]
@dotnet/roslyn-compiler for second review please

**Comment by chsienki** [comment]
Rebase onto master and fixed up the errors.

I also renamed `IteratorElementType` to `IteratorElementTypeWithAnnotations` to be consistent with @gafter's changes

**Pull request merged**